### PR TITLE
Reductions won't accept lists anymore

### PR DESCRIPTION
**Pull request by joaogui1**
Solves #3038 doing what was advised on #3050 by @jakevdp

### CODE REVIEW

**Comment by joaogui1** [comment]
So, I am allowing tuples as input, otherwise we will lose `jnp.prod(shape)`, which would be a shame

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -1488,7 +1489,10 @@ def reduction(a, axis=None, dtype=None, out=None, keepdims=False):
     if out is not None:
       raise ValueError("reduction does not support the `out` argument.")
 
-    a = a if isinstance(a, ndarray) else asarray(a)
+
+    if isinstance(a, list):
+      raise TypeError(f"Reductions don't work on list, pass a ndarray, tuple or scalar instead")
+    a = asarray(a)
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py:1495]: This change appears to be causing `masking_test.py` to fail.

(If nothing else you could kick the problem down the road by duplicating the logic as it was before the change.)

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -1488,7 +1489,10 @@ def reduction(a, axis=None, dtype=None, out=None, keepdims=False):
     if out is not None:
       raise ValueError("reduction does not support the `out` argument.")
 
-    a = a if isinstance(a, ndarray) else asarray(a)
+
+    if isinstance(a, list):
+      raise TypeError(f"Reductions don't work on list, pass a ndarray, tuple or scalar instead")
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py:1494]: Can we move this test into a helper function? We should be consistent about what we accept in most of these NumPy functions, not just reductions, and that is probably most easily achieved with a helper function that does the checking, e.g., `_check_arguments_are_arrays_or_scalars(...)` or similar?

Do you think we should accept tuples only in reductions? I'd lean towards not accepting them anywhere, so I'm curious whether you thing the `jnp.prod(shape)` pattern is common.

**Review by hawkinsp — CHANGES_REQUESTED** [review]

File: jax/numpy/lax_numpy.py (hunk 1 above)
Comment (hawkinsp) [reply jax/numpy/lax_numpy.py:1495]: Also note you are using an f-string here but aren't formatting anything.

File: jax/numpy/lax_numpy.py (hunk 2 above)
Comment (joaogui1) [reply jax/numpy/lax_numpy.py:1494]: It caused many errors, so it seems to be common in testing at least, but I need to check the logs to see where exactly they were being used.

File: jax/numpy/lax_numpy.py (hunk 1 above)
Comment (joaogui1) [reply jax/numpy/lax_numpy.py:1495]: On the f-string, the idea was to print a's type, then I removed, but maybe I should put it back, what do you think?
Also, I'm not exactly understanding why does it make masking_test fail, could you help me understand the error?
Thank you very much for the reviews :D

File: jax/numpy/lax_numpy.py (hunk 2 above)
Comment (shoyer) [reply jax/numpy/lax_numpy.py:1494]: I agree with @hawkinsp  that we should try to avoid both lists and tuples.

It might be cleaner to write a helper function like `_asarray_no_coercion()`, and just use that instead of `asarray()` internally.

Comment (joaogui1) [reply jax/numpy/lax_numpy.py:1494]: `host_callback_test`, `linalg_test` and `random_test` use this pattern, I could open a PR removing it, should I use `onp.prod` instead or do you guys prefer another fix?

Comment (shoyer) [reply jax/numpy/lax_numpy.py:1494]: `onp.prod(x)` is a good alternative. So is `functools.reduce(operator, x, 1)`, if a little more verbose.

Comment (joaogui1) [reply jax/numpy/lax_numpy.py]: I added the `_asarray_no_coercion()`, was this what you had in mind?

File: jax/numpy/lax_numpy.py (hunk 1 above)
Comment (joaogui1) [reply jax/numpy/lax_numpy.py]: > (If nothing else you could kick the problem down the road by duplicating the logic as it was before the change.)
Could you elaborate a little bit on this please?

**Comment by joaogui1** [comment]
So, right now it's failing on the masking test and parallel test, with the same error, can you guys help me out? @shoyer @hawkinsp 
Sorry for bothering

**Comment by jakevdp** [comment]
From the test failures, it looks like the tests are passing a list to a jax function that no longer accepts lists. One of the issues comes from the fact that reductions attempt to convert their inputs into arrays: https://github.com/google/jax/blob/fd886b17a48f23637f3fcae810117721f155e16c/jax/numpy/lax_numpy.py#L1520

But the test of reductions is passing a list: https://github.com/google/jax/blob/e36c72b9839efe8a98a75d93f87661f55b585ddb/jax/experimental/jax_to_tf/tests/tf_ops_test.py#L283-L284

**Comment by jakevdp** [comment]
A general comment: I'm a bit wary of this pull request as written. The problem is, it removes support for codepaths that are likely employed by many users (passing lists to jax functions), and so will almost certainly introduce breakages into user code.

For any change like this, we should be displaying a deprecation warning for a few releases: that is, if a user is passing input that currently works and will not work after this change, it should not break immediately, but rather raise a `FutureWarning` with instructions about how to fix the issue. Then, once that warning has been present for a few release cycles, we can make the breaking change.

**Comment by joaogui1** [comment]
Great idea Jake, this would be quite disruptive to codebases as it stands indeed, I opened a PR with the warning

**Comment by btanner** [comment]
Hey folks. I just encountered the future warning about reductions on list and tuples in the future. However, (and I only read quickly) - most of the chatter in here seemed to be about lists and excluded tuples from this change.

I just wanted to check in if this was intended. I've got a quite a few cute little tuple comprehensions that use jnp.sum and I'll rewrite them now (they are under review) if this is the direction JAX is taking. But if not, ..., well they are cute.

**Comment by jakevdp** [comment]
Hi @btanner - we are indeed moving to a world where tuples & lists will not be silently converted to device arrays by jax functions. If this is really what you want, you can always do it explicitly with something like `jnp.sum(jnp.asarray(tup))`. But if your tuples are small, the overhead of converting them to device arrays before summing is probably not what you want, and if your tuples are large enough that it is what you want, the overhead of constructing them as Python tuples (versus, say, native jax array operations) is probably not what you want.

**Comment by shoyer** [comment]
> Hi @btanner - we are indeed moving to a world where tuples & lists will not be silently converted to device arrays by jax functions. If this is really what you want, you can always do it explicitly with something like `jnp.sum(jnp.asarray(tup))`.

To be explicit, the other option (ideal for short tuples) looks like `functools.reduce(jnp.add, tup)`
